Quickstart (#8171)
mlflow/mlflow: docs/source/quickstart.rst

.. _quickstart:

Quickstart: Install MLflow, instrument code & view results in minutes
======================================================================

> In 15-20 minutes, you will:

Reviewer: 15-20 minutes sounds long. I think users can get through this in 10 minutes. How do you feel about saying "10 minutes"?

Author: Compromise "less than 15 minutes"? :-)

- Install MLflow
- Add MLflow tracking to your code

> - View run metrics in the MLflow tracking UI

Reviewer: ```suggestion
- View runs and experiments in the MLflow tracking UI
```

- (Optional) Run a tracking server to share results with others
- (Optional) Use Databricks to store your results
- Store the models produced by your runs

> - Run a model from a previous run

Reviewer: Nit: We typically refer to "model loading" and "model inference", rather than "running a model"

```suggestion
- Load a model from a previous run for inference
```

> As a data scientist, your explorations involve running your evolving training code many times. MLflow Tracking allows you to record important information your run, review and compare it with other runs, and share results with others. As an ML Engineer or MLOps professional, it allows you to compare, share, and deploy the best models produced by the team.

Reviewer: I love that we call out `As an ML Engineer or MLOps professional, it allows you to compare, share, and deploy the best models produced by the team.` at the end, but the quickstart doesn't cover these topics. We should definitely follow up with quickstart content for ML Engineers / MLOps. Just leaving this here as a note, no immediate action item

.. image:: _static/images/quickstart/quickstart_tracking_overview.png
    :width: 800px
    :align: center
    :alt: Diagram showing Data Science and MLOps workflow with MLflow

MLflow is available for Python, R, and Java, but this quickstart shows Python only. For Java, see :ref:`java_api`. For R, see :ref:`R-api`.

..
    Task: https://databricks.atlassian.net/browse/DOC-8674?atlOrigin=eyJpIjoiNjg5ZmYxODUzNjYxNDQzY2FjYTUxMTYyMDE4ZWFjNGQiLCJwIjoiaiJ9

Install MLflow
-----------------

> Install MLflow from PyPi using pip:

Reviewer: ```suggestion
Install MLflow from PyPI using pip:
```

Quickstart
==========

> Install MLFlow, instrument your code, and view a run's results in a few minutes:

Reviewer: ```suggestion
Install MLflow, instrument your code, and view a run's results in a few minutes:
```

How do I install MLFlow? 
------------------------

> You may install MLflow from PyPi using pip:

Reviewer: We probably want to mention that there's an R and Java library, but we don't need to show the examples in R or Java.

Author: Done in intro paragraphs and drill down.

.. code-section::

    .. code-block:: shell

        pip install mlflow

For more options, see tk.

> How do I add MLFlow tracking to my Python code?

Reviewer: Would also be super helpful to diagram or introduce the workflow that we'll go through in the quickstart. Data -> Model training -> Model selection & registration -> Deployment

> How do I add MLFlow tracking to my Python code?

Reviewer: Let's explicitly call out that the Data -> Training -> Registration is the data science workflow. Deployment is of interest to ML engineers / MLOps.

> How do I add MLFlow tracking to my Python code?

Author: Added para in intro, plus diagram

.. code-section::

    .. code-block:: python

        import mlflow

        from sklearn.model_selection import train_test_split
        from sklearn.datasets import load_diabetes
        from sklearn.ensemble import RandomForestRegressor

        mlflow.sklearn.autolog()

        db = load_diabetes()
        X_train, X_test, y_train, y_test = train_test_split(db.data, db.target)

        # Create and train models.
        rf = RandomForestRegressor(n_estimators=100, max_depth=6, max_features=3)
        rf.fit(X_train, y_train)

        # Use the model to make predictions on the test dataset.
        predictions = rf.predict(X_test)

In addition, or if you are using a library for which ``autolog`` is not yet supported, you may use key-value pairs to track:

.. list-table::
   :widths: 10 10 80 
   :header-rows: 1

   * - Name
     - Used for
     - Function call
   * - Parameters
     - Constant values (for instance, configuration parameters)
     - ``mlflow.log_param``
   * - Metrics
     - Values updated during the run (for instance, accuracy)
     - ``mlflow.log_metric``
   * - Artifacts
     - Files produced by the run (for instance, model weights)

>      - ``mlflow.log_artifacts``

Author: Done.

        import os
        from random import random, randint
        from mlflow import log_metric, log_param, log_artifacts

        if __name__ == "__main__":
            # Log a parameter (key-value pair)
            log_param("config_value", randint(0, 100))

            # Log a metric; metrics can be updated throughout the run
            log_metric("accuracy", random() / 2.0)
            log_metric("accuracy", random() + 0.1)
            log_metric("accuracy", random() + 0.2)

            # Log an artifact (output file)
            if not os.path.exists("outputs"):
                os.makedirs("outputs")
            with open("outputs/test.txt", "w") as f:
                f.write("hello world!")
            log_artifacts("outputs")

How do I view my MLFlow runs and experiments?
---------------------------------------------

Once you've run your code, you may view the results with MLFlow's tracking UI. To start the UI, run:

.. code-section::

    .. code-block:: shell

        mlflow ui

And then navigate to http://localhost:5000 in your browser. You will see a page similar to:

.. image:: _static/images/quickstart-ui-screenshot.png
    :width: 800px

>     :align: center

Reviewer: If possible, would be great to grab a screenshot of the new UI here

Author: New screenshot

> You are in the **Default** experiment, which now contains the tracking data for your run. You can view the parameters, metrics, and artifacts of your run by clicking on the run name. 

Reviewer: Maybe a short sentence following this explaining an experiment is a collection of runs. Can place it between 1 and 2

Author: Off-topic technical question: Is it best for an experiment to be a collection of runs with *identical* code (except for configuration values) or is it just "trying to get such-and-such a prediction from the domain, from day 1's terrible architecture to the super-bespoke architecture of day 1000"?

Reviewer: Great question! Definitely the latter - `"trying to get such-and-such a prediction from the domain, from day 1's terrible architecture to the super-bespoke architecture of day 1000"?`?

How do I share my MLFlow runs and experiments?
----------------------------------------------

> By default, MLFlow stores tracking data and artifacts in a ``./mlruns`` subdirectory of where you ran the code. You can change this behavior by setting the ``MLFLOW_TRACKING_URI`` environment variable to a different location. For instance, you can set it to a shared filesystem, a SQLAlchemy-compatible database, a tracking server, or a Databricks workspace.

Reviewer: Typically, to change the location, you would use `mlflow.set_tracking_uri()` from Python. The environment variable is helpful for provisioning a shared image or environment where data scientists (DS) can write their code without reasoning about the tracking URI.

Author: Rewrote this H2 to be API-centric

> By default, MLFlow stores tracking data and artifacts in a ``./mlruns`` subdirectory of where you ran the code. You can change this behavior by setting the ``MLFLOW_TRACKING_URI`` environment variable to a different location. For instance, you can set it to a shared filesystem, a SQLAlchemy-compatible database, a tracking server, or a Databricks workspace.

Reviewer: Would also be super helpful when we say that the data is stored in `./mlruns` by default to explain what `mlflow ui` is doing. Basically, it visualizes the contents of that local filesystem directory.

Author: Done

> You can run a tracking server on a network-accessible server by running:

Reviewer: Would be great to explain what a tracking server is and how it differs from the storage in `./mlruns`.

Author: Added this in the first para of the "sharing" H2

.. code-section::

    .. code-block:: shell

        mlflow server

On your development machine, set the ``MLFLOW_TRACKING_URI`` environment variable to the URL of that server. For example, if your tracking server is listening on the default port of **5000** on the machine with IP address **192.168.0.1**, you can set the ``MLFLOW_TRACKING_URI`` environment variable to:

.. code-section::

    .. code-block:: shell

        export MLFLOW_TRACKING_URI=http://192.168.0.1:5000

Now, when you run your code, it will send tracking data to the tracking server. To view the results, run the tracking UI on your local machine and point it to the tracking server:

.. code-section::

    .. code-block:: shell

>         mlflow ui --backend-store-uri http://192.168.0.1:5000

Reviewer: Should just be able to navigate to `http://192.168.0.1:5000` to view the UI

Author: Removed this code and changed text to "browse to tracking server"

..

    CONTENT REVIEW: Maybe that's all we need for *this* page? Short and sweet, you've: 

    - installed MLFlow
    - used `autolog` and explicit logging
    - run the tracking UI
    - run the tracking server

    Maybe we end there and "Next Steps" it? 

    If not:

How do I use MLFlow with a Databricks workspace?
------------------------------------------------

> tk 

Reviewer: Would be great to link to the Databricks docs for this; @dbczumar to find the relevant docs link.

Author: I link to the Databricks QS.

> 

Reviewer: Would be awesome to cover the model selection and production aspects. Show the user how to use the UI to pick a model, register with the MLflow Model Registry, and then deploy with `mlflow models serve` (local serving). Should link to the full MLflow Models & Model Registry references for more info about moving to production / production lifecycle.

Author: I am worried about scope creep. I think there's a logical follow-on doc that's a quickstart for the MLOps-y review, validate, serve, monitor stuff, but I don't want to make the follow-along time for this page much >20 minutes.

> How do I store a model in MLFlow?

Reviewer: Autologging fortunately already stored, but would be great to indicate that mlflow.{flavor_name}.log_model() can be used to log models.

Author: Wrote H2 and sample

---------------------------------

tk 

How do I run a model artifact from a specific MLFlow run?
---------------------------------------------------------

tk

> How do I do a hyperparameter sweep with MLFlow Tracking?

Reviewer: Can probably leave this out for now, but it would be great to call out Projects and Recipes briefly at the end of this workflow. "For more information about creating reproducible model training workflows, check out MLflow Projects.", "For prepackaged training and deployment solutions for classification and regression modeling, check out MLflow Recipes"

Author: Removed this H2.

--------------------------------------------------------

tk

Next Steps
----------
.. 
    First, code:

- Example: MLFlow tutorial
- MLFlow example directory

.. 

>     Drilldown for each H2. These could also be at the bottom

Reviewer: Would be great to include the drilldown in the H2s of the relevant sections

    of their respective sections. Depends on info architecture design.

- Customizing and troubleshooting MLFlow installation
- Using Autolog, including a list of what libraries are supported
- Effectively using the MLFlow Tracking API
- Choosing and configuring the MLFlow tracking backend 
- Navigating the MLFlow Tracking UI

.. 
    Next tool: either Models or Projects

- Storing models with MLFlow tracking
- End-to-end tracking, collaboration, and deployment with MLFlow Projects

.. 
    More top-down-y stuff

- Understanding MLFlow's concepts, tools, and APIs 
- Why MLFlow?
  

{>> Important things in original we don't want to lose:

- mlflow skinny and extras and why you might choose one or t'other

> - Do we document / walkthrough virtualenvs or, e.g., conda or is that all assumed?

Reviewer: We should use the following for model serving because it doesn't depend on conda or virtualenv.

`mlflow models serve --env-manager local` 

Can then say that there's a way to restore the model's environment for serving and link to the MLflow Models page. Don't need to go into detail about conda / virtualenv here.

Author: Punted this over to the drill down page

> - What about R? 

Reviewer: Just calling out that the library can be installed in Java / R is plenty here. R isn't widely used.

Author: Done in the drill down. I actually ended up linking over to the R API page, since there's an intro page. For Java, I "due to the complexity of the Java ecosystem..." bailed.

> - Mac python on install (relates to env answer)

Reviewer: Keeping the mac note seems helpful, since macs are pretty ubiquitous in industry at this point

Author: Done in the drill down

- repo contains a big helpful directory of examples 

> - WORKER_TIMEOUT issue 

Reviewer: Don't have to worry about this for quickstart

Author: Deleted. Haven't re-added in the drill down

        if __name__ == "__main__":
            # Log a parameter (key-value pair)

>             log_param("config_value", randint(0, 100))

Reviewer 2: it might be nice to show the dict-based logging here as well

```
# Log a collection of parameters
run_params = {
  "optimizer": "sgd",
  "seed": 98765,
  "n_iterations": 2500,
}
mlflow.log_params(run_params)
```

Author: TODO

                f.write("hello world!")
            log_artifacts("outputs")

- For more details on autolog, including a list of what libraries are supported, see :ref:`quickstart_drilldown_autolog`.

> - For additional functions such as ``log_text`` and ``log_image``, see :ref:`quickstart_drilldown_tracking_api`.

Author: Done.

Viewing MLflow runs and experiments
===================================

> Once you've run your code, you may view the results with MLflow's tracking UI. To start the UI, run:

Author: I've done this in the opening sentences of the "sharing" H2. Not super-happy with the flow of the text, but it covers the bases.

.. code-section::

    .. code-block:: shell

        databricks configure

Set the ``Databricks Host`` to the URL of your Databricks workspace, and set the ``Username`` and ``Password`` to the credentials you use to access the workspace. If you've created an authentication token for your Databricks workspace (``databricks tokens create``), you can use it instead of your password. Call ``databricks configure`` with the ``-t, \--token`` option.

..
    Consider adding a link to the Databricks CLI docs for tokens create

In your training code, modify the call to ``mlflow.set_tracking_uri`` to use Databricks and set the experiment to the path of your experiment in Databricks, replacing ``user_name`` and ``experiment_name`` with the appropriate values:

.. code-section::

    .. code-block:: python

        mlflow.set_tracking_uri("databricks")
        mlflow.set_experiment(f"/Users/{user_name}/{experiment_name}")

If the specified experiment does not exist, it will be created.

Storing a model in MLflow
=========================

An MLflow Model is a directory that packages machine learning models and support files in a standard format. The directory contains:

> - An **MLModel** file in YAML format specifying the model's **flavor** (or **flavors**);

Reviewer 2: ```suggestion
- An **MLModel** file in YAML format specifying the model's **flavor** (or **flavors**), dependencies, signature (if supplied), and important metadata;
```

Author: Added

- The various files required by the model's flavor(s) to instantiate the model. This will often be a serialized Python object; and

> - Files necessary for recreating the model's runtime environment (for instance, a **conda.yaml** file)

Reviewer 2: Would it be helpful to show a screenshot of the contents of an MLModel file and the supplementary saved objects? 
![Screenshot 2023-04-12 at 9 48 49 AM](https://user-images.githubusercontent.com/39283302/231478456-27873965-80a7-47f3-877d-f5ebd688e121.png)

> - Files necessary for recreating the model's runtime environment (for instance, a **conda.yaml** file)

Author: I think this should be in the drill down coverage

> - Files necessary for recreating the model's runtime environment (for instance, a **conda.yaml** file)

Author: I added the "input example if specified"

.. image:: _static/images/quickstart_run_id.png
    :width: 400px
    :align: center
    :alt: Screenshot showing location of run ID in tracking UI

.. code-section::

    .. code-block:: python

        import mlflow

        from sklearn.model_selection import train_test_split
        from sklearn.datasets import load_diabetes

        db = load_diabetes()
        X_train, X_test, y_train, y_test = train_test_split(db.data, db.target)

        model = mlflow.sklearn.load_model("runs:/97f4e98c5f3645c8acd800cbddf5f6da/model")
        predictions = model.predict(X_test)
        print(predictions)

Note that while ``log_model`` saves environment-specifying files such as **conda.yaml** and **requirements.txt**, ``load_model`` does not automatically recreate that environment. To do so, you need to use your preferred method (**conda**, **virtualenv**, **pip**, etc.), using the artifacts saved by ``log_model``.

To learn more about loading runs, see :ref:`quickstart_drilldown_log_and_load_model`.

> 

Reviewer 2: It might be nice to show the programmatic access pattern for retrieving the model_uri via the run context:

```
with mlflow.start_run() as run:
  mlflow.sklearn.log_model(
    my_model,
    artifact_path="model",
  )

model_uri = f"runs:/{run.info.run_id}/model"
loaded_model = mlflow.sklearn.load_model(model_uri)
```

Author: TODO

                f.write("hello world!")
            log_artifacts("outputs")

If you are using a library that supports autologging, but wish to disable it, you may do so by calling ``mlflow.autolog(disable=True)``.

- For more details on automatic logging, see :ref:`automatic-logging`.
- For more details on the explicit logging API, see :ref:`tracking_logging_functions`.

View MLflow runs and experiments
-----------------------------------

Once you've run your code, you may view the results with MLflow's tracking UI. To start the UI, run:

.. code-section::

    .. code-block:: shell

        mlflow ui

And then navigate to http://localhost:5000 in your browser. You will see a page similar to:

> .. image:: _static/images/quickstart/quickstart_ui_screenshot.png

Reviewer: Thanks for updating the screenshot! A couple small asks:

1. Can we show some metrics and parameters in the "Table view" screenshot? Right now, it's hard to see that MLflow recorded these.

2. Can we include a second screenshot showcasing the "Chart view" with a couple of charts?

Author: For both table and chart, the top-level screen shows the runs, not the metrics/params.

So I need to drill down, which is fine, but: 

For the RandomForestRegressor sample, it does it all in a single pass, so the metrics don't go up over time. So I need to use a screenshot from a different sample, which is fine, but I'll need to write around it a little.

    :width: 800px
    :align: center
    :alt: Screenshot of MLflow tracking UI
..

You are in the **Default** experiment, which now contains the tracking data for your run. An experiment is a collection of related runs. You can view the parameters, metrics, and artifacts of your run by clicking on the run name.

For more details on the tracking UI, see :ref:`tracking`.

.. _quickstart_tracking_server:

Share MLflow runs and experiments
-----------------------------------

For getting started, the last example stored the tracking data locally. Generally, you will want to use shared storage. Locally, MLflow stores tracking data and artifacts in an **mlruns/** subdirectory of where you ran the code. The tracking UI, when run locally, visualizes this. You can store your data remotely by:

- Calling ``mlflow.set_tracking_uri`` in your code; or
- Setting the ``MLFLOW_TRACKING_URI`` environment variable

> With either option, you can track your runs in a shared filesystem, a SQLAlchemy-compatible database, a tracking server, or a Databricks workspace.

Reviewer: Love that we're calling this out! I found it a bit confusing that we're telling people to call this Python API / set this environment variable before we tell them about the possible values they can use (file system URI for shared filesystem, SQLAlchemy URI for database, HTTP/S URL for tracking server, etc).

Perhaps we should move this below the discussion of "what is a tracking server?". It would also be great to include a couple of examples showing how you can use `set_tracking_uri` and `MLFLOW_TRACKING_URI` with a shared filesystem and with a SQL database. Can we also link to the relevant MLflow Tracking docs so that users can find more info?

A tracking server is a lightweight HTTP server. You can run a tracking server on a network-accessible server by running:

.. code-section::

    .. code-block:: shell

        mlflow server

For instance, if you've run the above command on a machine with IP address **192.168.0.1** and port **5000**, you can add tracking data to it either by:

.. code-section::

    .. code-block:: python

        mlflow.set_tracking_uri("http://192.168.0.1:5000")
        mlflow.autolog()  # Or other tracking functions

Or, on your development machine, you can set the ``MLFLOW_TRACKING_URI`` environment variable to the URL of that server:

.. code-section::

    .. code-block:: shell

        export MLFLOW_TRACKING_URI=http://192.168.0.1:5000

Now, when you run your code, it will send tracking data to the tracking server. You can view the tracking data by navigating to the URI with a browser.

There are many options available for the tracking backend. For more details, see :ref:`tracking_server`.

> Use MLflow with a Databricks workspace

Reviewer: I think we should make this a subheader of "Share MLflow Runs and experiments"

Author: ~~If we do this, it will drop off the TOC. ~~ DONE

> ----------------------------------------

Author: DONE

..
    Consider adding a link to the Databricks CLI docs for tokens create

In your training code, modify the call to ``mlflow.set_tracking_uri`` to use Databricks and set the experiment to the path of your experiment in Databricks, replacing ``user_name`` and ``experiment_name`` with the appropriate values:

.. code-section::

    .. code-block:: python

        mlflow.set_tracking_uri("databricks")
        mlflow.set_experiment(f"/Users/{user_name}/{experiment_name}")

If the specified experiment does not exist, it will be created.

Store a model in MLflow
-------------------------

An MLflow Model is a directory that packages machine learning models and support files in a standard format. The directory contains:

- An **MLModel** file in YAML format specifying the model's **flavor** (or **flavors**), dependencies, signature (if supplied), and important metadata;
- The various files required by the model's flavor(s) to instantiate the model. This will often be a serialized Python object;
- Files necessary for recreating the model's runtime environment (for instance, a **conda.yaml** file); and

> - Optionally, an input sample

Reviewer: Nit: "Input example" is used throughout the docs and in the API

```suggestion
- Optionally, an input example
```

> When using autologging, MLflow will automatically log the run's model. You can also log a model manually by calling ``mlflow.{library_module_name}.log_model``. In addition, if you wish to load the model soon, it may be convenient to output the run's ID directly to the console. For that, you'll need the object of type ``mlflow.ActiveRun`` for the current run. You get that object by wrapping all of your logging code in a ``with mlflow.start_run() as run:`` block. For example:

Reviewer: Nit: I think it's a bit unclear what "run's model" means; a run could have multiple models or no models at all

```suggestion
When using autologging, MLflow will automatically log models to the runs that it creates. You can also log a model manually by calling ``mlflow.{library_module_name}.log_model``. In addition, if you wish to load the model soon, it may be convenient to output the run's ID directly to the console. For that, you'll need the object of type ``mlflow.ActiveRun`` for the current run. You get that object by wrapping all of your logging code in a ``with mlflow.start_run() as run:`` block. For example:
```

Reviewer: Nit: Can we link to the Python API docs for `start_run()` and `ActiveRun`?

Author: DONE

            X_train, X_test, y_train, y_test = train_test_split(db.data, db.target)

            # Create and train models.
            rf = RandomForestRegressor(n_estimators=100, max_depth=6, max_features=3)
            rf.fit(X_train, y_train)

            # Use the model to make predictions on the test dataset.
            predictions = rf.predict(X_test)
            print(predictions)

            mlflow.sklearn.log_model(rf, "model")

            print("Run ID: {}".format(run.info.run_id))

In the case of the ``sklearn`` flavor, ``log_model`` stores the following files in the **artifacts** directory of the run's directory on the tracking server:

.. code-section::

    .. code-block:: shell

        model/
        |-- MLmodel
        |-- conda.yaml
        |-- model.pkl
        |-- python_env.yaml
        |-- requirements.txt

If you've not called ``set_tracking_uri`` or set the ``MLFLOW_TRACKING_URI`` environment variable to point to a remote tracking server, this **model** directory will be under the **mlruns** directory.

For more information, including a list of supported model flavors and storing your own flavor, see :ref:`models_built-in-model-flavors`.

> Load a model from a specific training run for inference
> --------------------------------------------

Reviewer 2: ```suggestion
Load a model from a specific training run for inference
------------------------------------------------------
```
.rst linting is complaining about this. We have lint checks enabled that mandate the section delimiter must be at least as long as the title.

To load and run a model stored in a previous run, you can use the ``mlflow.{library_module_name}.load_model`` function. You'll need the run ID of the run that logged the model. You can find the run ID in the tracking UI:

.. image:: _static/images/quickstart/quickstart_run_id.png
    :width: 400px
    :align: center
    :alt: Screenshot showing location of run ID in tracking UI

.. code-section::

    .. code-block:: python

        import mlflow

        from sklearn.model_selection import train_test_split
        from sklearn.datasets import load_diabetes

        db = load_diabetes()
        X_train, X_test, y_train, y_test = train_test_split(db.data, db.target)

        model = mlflow.sklearn.load_model("runs:/97fd7ade5106ee341e0b4c63a53a9776231")
        predictions = model.predict(X_test)
        print(predictions)

> Note that while ``log_model`` saves environment-specifying files such as **conda.yaml** and **requirements.txt**, ``load_model`` does not automatically recreate that environment. To do so, you need to use your preferred method (**conda**, **virtualenv**, **pip**, etc.), using the artifacts saved by ``log_model``.

Reviewer: Great call-out! Can we also state that MLflow makes this easy to do? Wouldn't want users to worry that they have to set up the conda / virtualenv by themselves

Author: DONE

To learn more about loading models for specific runs, see :ref:`quickstart_drilldown_log_and_load_model`.

Next steps
----------
..
    First, code:

- :ref:`MLflow tutorials and examples <tutorials-and-examples>`
- Use the MLflow Registry to store and share versioned models, see :ref:`registry`
- Use MLflow Projects for packaging your code in a reproducible and reusable way, see :ref:`projects`

> - Use MLflow Recipes (experimental) to create workflows for faster iterations and easier deployment, see :ref:`recipes`

Reviewer 2: We can remove "experimental" references for MLflow Recipes. It's been stabilized with the 2.3.0 release

Author: DONE

- Set the ``MLFLOW_TRACKING_URI`` environment variable

A tracking server is a lightweight HTTP server built in to MLflow. You can run a tracking server on a network-accessible server by running:

.. code-section::

    .. code-block:: shell

        mlflow server

For instance, if you've run the above command on a machine with IP address **192.168.0.1** and port **5000**, you can add tracking data to it either by:

.. code-section::

    .. code-block:: python

        mlflow.set_tracking_uri("http://192.168.0.1:5000")
        mlflow.autolog()  # Or other tracking functions

Or, on your development machine, you can set the ``MLFLOW_TRACKING_URI`` environment variable to the URL of that server:

.. code-section::

    .. code-block:: shell

        export MLFLOW_TRACKING_URI=http://192.168.0.1:5000

Now, when you run your code, it will send tracking data to the tracking server. You can view the tracking data by navigating to the URI with a browser.

There are many options available for the tracking backend. For more details, see :ref:`tracking_server`.

Use MLflow with a Databricks workspace
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> You need to configure Mlflow to use your Databricks workspace (To get started with Databricks, see: `Get started: Account and Workspace setup <https://docs.databricks.com/getting-started/index.html>`_). You will need to know the URL of your Databricks workspace. You can find the URL in the Configuration page of the workspace:

Reviewer: ```suggestion
You need to configure MLflow to use your Databricks workspace (To get started with Databricks, see: `Get started: Account and Workspace setup <https://docs.databricks.com/getting-started/index.html>`_). You will need to know the URL of your Databricks workspace. You can find the URL in the Configuration page of the workspace:
```

        |-- python_env.yaml
        |-- requirements.txt

If you've not called ``set_tracking_uri`` or set the ``MLFLOW_TRACKING_URI`` environment variable to point to a remote tracking server, this **model** directory will be under the **mlruns** directory.

For more information, including a list of supported model flavors and storing your own flavor, see :ref:`models_built-in-model-flavors`.

Load a model from a specific training run for inference
--------------------------------------------------------

To load and run a model stored in a previous run, you can use the ``mlflow.{library_module_name}.load_model`` function. You'll need the run ID of the run that logged the model. You can find the run ID in the tracking UI:

.. image:: _static/images/quickstart/quickstart_run_id.png
    :width: 400px
    :align: center
    :alt: Screenshot showing location of run ID in tracking UI

.. code-section::

    .. code-block:: python

        import mlflow

        from sklearn.model_selection import train_test_split
        from sklearn.datasets import load_diabetes

        db = load_diabetes()
        X_train, X_test, y_train, y_test = train_test_split(db.data, db.target)

>         model = mlflow.sklearn.load_model("runs:/97fd7ade5106ee341e0b4c63a53a9776231")

Reviewer: Can we use the run ID from the screenshot above for consistency?

        predictions = model.predict(X_test)
        print(predictions)

> Note that while ``log_model`` saves environment-specifying files such as **conda.yaml** and **requirements.txt**, ``load_model`` does not automatically recreate that environment. To do so, you need to use your preferred method (**conda**, **virtualenv**, **pip**, etc.), using the artifacts saved by ``log_model``. If you serve your model with ``mlflow models serve`` or ``mlflow run``, MLflow will automatically recreate the environment. Those commands also accept an ``--env-manager`` option for even more control. (This is described in detail in :ref:`model-enviroment-management`.)

Reviewer: Nit: `mlflow run` isn't used for model serving. 

```suggestion
Note that while ``log_model`` saves environment-specifying files such as **conda.yaml** and **requirements.txt**, ``load_model`` does not automatically recreate that environment. To do so, you need to use your preferred method (**conda**, **virtualenv**, **pip**, etc.), using the artifacts saved by ``log_model``. If you serve your model with ``mlflow models serve``, MLflow will automatically recreate the environment. Those commands also accept an ``--env-manager`` option for even more control. (This is described in detail in :ref:`model-enviroment-management`.)
```

Can we also mention that ``mlflow.pyfunc.spark_udf()`` supports an `env_manager` flag to recreate the model environment during batch inference with Spark? It's set to ``local`` by default but can be set to ``virtualenv`` and ``conda``.
